yarn-project/simulator/src/public/enqueued_call_side_effect_trace.ts

```diff
@@ -388,6 +399,25 @@ export class PublicEnqueuedCallSideEffectTrace implements PublicSideEffectTraceI
     lowLeafIndex: Fr = Fr.zero(),
     lowLeafPath: Fr[] = emptyNullifierPath(),
   ) {
+    // We have hinted this bytecode, we do nothing.
+    if (
+      this.previousBytecodeHints.has(contractInstance.contractClassId.toString()) ||
+      this.avmCircuitHints.contractBytecodeHints.has(contractInstance.contractClassId.toString())
+    ) {
+      this.log.debug(
+        `Contract class id ${contractInstance.contractClassId.toString()} already exists in previous hints`,
+      );
+      return;
+    }
+
+    // Before adding this hint, check that we won't be exceeding the MAX_UNIQUE_CONTRACTS_PER_TX
+    if (
+      this.previousBytecodeHints.size + this.avmCircuitHints.contractBytecodeHints.size >=
+      AVM_MAX_UNIQUE_CONTRACT_CALLS
+    ) {
+      throw new UniqueContractCallsLimitReachedError(AVM_MAX_UNIQUE_CONTRACT_CALLS);
+    }
+
```

Comment:
This isn't as bad as I thought it'd be. If you just always avoid inserting duplicates into a forked trace, then this is actually quite clean.

But this doesn't work if you for a fork, right? The access to `previousBytecodeHints` needs to recurse to its grandparent.

This is what ended up leading me to the PublicStorage and Nullifiers classes. You can just have a wrapper class that contains the current call's map, the parent, a getter function that falls back on a recursion into the parent, and a setter function that checks by recursing into the parent.